src/drivers/uavcan/uavcan_main.cpp

```diff
@@ -712,6 +722,9 @@ UavcanNode::Run()
 			if (request.message_type == uavcan_parameter_request_s::MESSAGE_TYPE_PARAM_REQUEST_READ) {
 				uavcan::protocol::param::GetSet::Request req;
 
+				// sleep a bit to reduce spike on CAN bus traffic
+				usleep(100);
```

Comment:
This may sleep up to one systick, in our case 10ms , i.e. not "a bit". Is this intended?

Reply:
Thanks for pointing out this. I didnt realize that this is not simply 100us. Anyway, this sleep only happens when there is uORB param request from QGC (new QGC from newly connected GS mostly since restarted QGC from already connected GS would read params again from cache). For preflight startup when QGC is connected , I think this shouldnt create any performance setback. But during mid-fly, if a new QGC is connected and request params, this sleep would affect the ESC control. the ESC control loop need to publish cmd at 400Hz aka every 2.5ms. I just looked into the config CONFIG_USEC_PER_TICK. It is 1000us , same for saluki and other platforms (https://github.com/tiiuae/saluki-v1/blob/ab86c9c5a7d50b3334e38163d9d7f50d46315790/nuttx-config/bootloader/defconfig#L101 ) . So I assume if I use usleep(100) then it might sleep up to 1 tick aka 1ms < 2.5ms which is acceptable, right? But where is it defined 10ms as you stated?

Reply:
No, actually we are using 10000 systick, not 1000 like upstream px4

Reply:
So the Ŕun method is only being called every 10ms anyway. I don't see why you need sleep here? Why does it cause spike on can bus traffic if there is no sleep?

Reply:
I can explain the problem I encountered previously during CAN flight test. It doesn't happen every single time but when it does, during QGC startup upon a new connection, it needs to retrieve the drone PX4 parameters as well as all the UAVCAN parameters from all detected UAVCAN nodes. This leads to a huge spike (due to requests and response msgs) in UAVCAN traffic usually on the background of around 400Hz of ESC cmd msgs published. this leads to some of the retrieved params being corrupted or many are not retrieved successfully and QGC doesn't start up properly. I have tried to do add some resend request functionalities if parameter GetSet fails but that is not enough to fix the whole issue. Reducing the parameter request rate by sleeps was the only solution that makes QGC able to retrieve all the params and start up successfully. 
Or maybe, I can limit the publish rate of topic ORB/topics/uavcan_parameter_request https://github.com/tiiuae/px4-firmware/blob/b490068b35f7e5b977415c75c1b34a054fb20b11/src/drivers/uavcan/uavcan_main.hpp#L268  ?


![image](https://user-images.githubusercontent.com/17795294/216256432-cd870812-a845-478e-9049-02b0a7592d86.png)

Reply:
btw, this is test at classical CAN maximum bitrate of 1M. This means theoretically, the frame throughput should be 7633 FPS. But sometimes, if FPS > 1000, there might be chance of corrupted frames.

Reply:
Understood!

So basically the core reason of why the frames get corrupted is not well known, and you have found that just reducing the rate helps.

Can you try if you can limit the publish rate of topic ORB/topics/uavcan_parameter_request ? That *should* resolve the issue I think?

I think it is safer to avoid sleeping in the "Run", if possible

Reply:
or i can try using <uORB::SubscriptionInterval>	uORB::Subscription _param_request_sub{ORB_ID(uavcan_parameter_request)} --- > uORB::SubscriptionInterval _param_request_sub{ORB_ID(uavcan_parameter_request), micro_second_interval}
micro_second_interval can be adjusted during test so that no message in publish queue is lost?

Reply:
or i can try using <uORB::SubscriptionInterval>	uORB::Subscription _param_request_sub{ORB_ID(uavcan_parameter_request)} --- > uORB::SubscriptionInterval _param_request_sub{ORB_ID(uavcan_parameter_request), micro_second_interval}
micro_second_interval can be adjusted during test so that no message in publish queue is lost?

Reply:
I think you can safely set the interval there, and still won't loose parameter requests. I don't think that the mavlink would publish the requests until it gets response to the previous. I might be wrong, but this is probably the right way to fix it.

Reply:
yeah, i will try this out. Next CAN drone test, we will see if this also solves the problem